### DISCUSSION TITLE
Catastrophic bug that led me to lose months of work

### DISCUSSION THREAD

**Discussion by HenriDeh** [community/community#24422] [Apps, API and Webhooks]
<p>I encountered a bug that made me lose tons of work (see <a href="https://github.community/t/git-overwrote-and-mixed-up-my-files/2083">this thread</a>). I am new to git, but this is probably not a desired behavior, espacially that the manipulation I did is typically something you would do when you discover Git. I did it using Atom so I will describe the steps to reproduce it on Atom. However, I don’t think that’s related to Atom. Here is how to reproduce it:</p>
<p>Step 1. Create a new project in Atom (My Project) </p>
<p>Step 2. In the project create a folder (Folder 1)</p>
<p>Step 3. In that project create a source file (aFile.jl) and write “This is the file in folder 1” in it, save</p>
<p>Step 4. Duplicate Folder 1 twice, name the new folders Folder 2 and Folder 3 respectively.</p>
<p>Step 5. Change the text in Folder 2/aFile.jl to “This is the file in folder 2”, save. Do the same in folder 3 but with a 3.</p>
<p>Step 6. In the Git tab, initialize a new repository. Your project tree should now look like this:</p>
<p>&gt; My Project</p>
<p>    &gt; .git</p>
<p>    &gt; Folder 1</p>
<p>        - aFile.jl</p>
<p>    &gt; Folder 2</p>
<p>        - aFile.jl</p>
<p>    &gt; Folder 3</p>
<p>        - aFile.jl</p>
<p>Step 7. In the Git tab, the three files should be awaiting in <em>unstaged changes</em>. Click on the … next to it and _discard all changes. _If you now check the tree the “aFile.jl” files disapeared. </p>
<p>Step 8. In the Git tab, click on the same … and <em>undo last discard</em>. This will bring back the files.</p>
<p>However, if you check their contents, the “aFile.jl” files now have the same content. Which means that the undo pick a single file and wrote it in every folder. I know that Step 7. is stupid, but I did it thinking Git was “safe” for my files. This is not an expected behaviour. If this is not a bug and you know a way of restoring the three files in the state they were at Step 6. I would be eternally grateful for a solution. If it is a bug, then someone may want to fix that. </p>

**Comment by lee-dohm** [comment, marked as answer]
<p>Yes, this is a bug. Please open an issue at <a href="https://github.com/atom/github/issues/new" rel="nofollow noopener">https://github.com/atom/github/issues/new</a> with the same information and we’ll look into it.</p>
<p>Unfortunately, this has nothing to do with git. Because the changes were unstaged, git was never tracking those files and therefore wasn’t protecting them. For future reference, in order for git to protect your files or their contents, you have to:</p>
<ol>
<li>Stage the files (or their latest contents)</li>
<li>Commit the changes</li>
</ol>
<p>Typically, I do the above multiple times a day so that I only lose at most a couple hours of work.</p>
<p>Additionally, if you want to learn more about exactly how git works and protects your files, I highly recommend <a href="https://jwiegley.github.io/git-from-the-bottom-up/" rel="nofollow noopener">Git from the Bottom Up</a> by John Wiegley.</p>
<p>But you are correct, the undo discard feature in Atom’s github package shouldn’t have behaved that way.</p>
